Features/178 tile

ClaudiaComito opened the pull request:
## Description

<!--- Include a summary of the change/s.
Please also include relevant motivation and context. List any dependencies that are required for this change.
--->
Implementation of heat version of np.tile():
https://numpy.org/doc/stable/reference/generated/numpy.tile.html

- process-local operations via torch.repeat()
- distributed tiling via `Alltoallv`
~~- also, introducing sanitation module #468 (with only a few basic functions), the idea is to expand it on the fly as needed~~ 

Issue/s resolved: #178 

## Changes proposed:
- Implemented new function: `manipulations.tile()`, after https://numpy.org/doc/stable/reference/generated/numpy.tile.html


## Type of change
<!--
i.e.
- Bug fix (non-breaking change which fixes an issue)
- New feature (non-breaking change which adds functionality)
- Breaking change (fix or feature that would cause existing functionality to not work as expected)
- Documentation update
--->
- New feature (non-breaking change which adds functionality)

## Due Diligence

- [x] All split configurations tested
- [x] Multiple dtypes tested in relevant functions
- [x] Documentation updated (if needed)
- [x] Updated changelog.md under the title "Pending Additions"

#### Does this change modify the behaviour of other functions? If so, which?
no

mtar commented:
Is the change of the split axis intended?

```python
>>> ht.tile(ht.array([1,2,3],split=0), [1,1])
DNDarray([[1, 2, 3]], dtype=ht.int64, device=cpu:0, split=1)
```

ClaudiaComito commented:
> Is the change of the split axis intended?
> 
> ```python
> >>> ht.tile(ht.array([1,2,3],split=0), [1,1])
> DNDarray([[1, 2, 3]], dtype=ht.int64, device=cpu:0, split=1)
> ```

@mtar belated thanks for commenting, yes you are tiling a 1-D array with  2-dimensional repetitions, in this case `tile()` will prepend the missing dimension (just like `numpy.tile()`, so the split axis nominally increases by one in this case.

coquelin77 commented:
rerun tests

coquelin77 commented on heat/core/manipulations.py:
```diff
@@ -3254,6 +3255,235 @@ def vstack(arrays: Sequence[DNDarray, ...]) -> DNDarray:
+            comm=x.comm,
+            balanced=x.balanced,
+        )
+    else:
```
very very minor comment: since you have a return in the line above, you dont need an else block here at all

coquelin77 reviewed (changes requested):
the code itself looks fine, but i found a bug in the alltoallv logic:

```python
x = ht.arange(3*3).reshape((3, 3)).resplit(1)
print(x)
>>> DNDarray([[0, 1, 2],
              [3, 4, 5],
              [6, 7, 8]], dtype=ht.int32, device=cpu:0, split=1)
y = ht.tile(x, (2, 3))
print(y)
>>> DNDarray([[        0,         4,         2,         0,         4,         2,         0,         4,         2],
              [        1,         1,         0,         1,     22065,     32628,         1,     32628,         0],
              [        3,         0,         0,         3,         0, 724118576,         3,         0,         0],
              [        0,         4,         2,         0,         4,         2,         0,         4,         2],
              [        1,         1,         0,         1,     22065,     32628,         1,     32628,         0],
              [        3,         0,         0,         3,         0, 724118576,         3,         0,         0]], dtype=ht.int32, device=cpu:0, split=1)
```

coquelin77 commented:
> > the code itself looks fine, but i found a bug in the alltoallv logic:
> > ```python
> > x = ht.arange(3*3).reshape((3, 3)).resplit(1)
> > print(x)
> > >>> DNDarray([[0, 1, 2],
> >               [3, 4, 5],
> >               [6, 7, 8]], dtype=ht.int32, device=cpu:0, split=1)
> > y = ht.tile(x, (2, 3))
> > print(y)
> > >>> DNDarray([[        0,         4,         2,         0,         4,         2,         0,         4,         2],
> >               [        1,         1,         0,         1,     22065,     32628,         1,     32628,         0],
> >               [        3,         0,         0,         3,         0, 724118576,         3,         0,         0],
> >               [        0,         4,         2,         0,         4,         2,         0,         4,         2],
> >               [        1,         1,         0,         1,     22065,     32628,         1,     32628,         0],
> >               [        3,         0,         0,         3,         0, 724118576,         3,         0,         0]], dtype=ht.int32, device=cpu:0, split=1)
> > ```
> 
> Good catch! Fixed

then this is ready as far as i can tell
